File: src/main/java/work/nekow/particledrawing/core/network/ParticleSpawnPayload.java
package work.nekow.particledrawing.core.network;

import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.Identifier;
import net.minecraft.world.phys.Vec3;
import work.nekow.particledrawing.api.Color;
import work.nekow.particledrawing.api.ParticleStyle;

import java.util.UUID;

@SuppressWarnings("unused")
public record ParticleSpawnPayload(
    UUID particleId,
    ParticleStyle style,
    double x, double y, double z,
    float r, float g, float b, float a,
    float scale,
    int lifetime,
    UUID groupId,
    boolean glowing
) implements CustomPacketPayload {

    public static final Type<ParticleSpawnPayload> TYPE = new Type<>(
        Identifier.fromNamespaceAndPath("particledrawing", "particle_spawn")
    );

    public static final StreamCodec<FriendlyByteBuf, ParticleSpawnPayload> STREAM_CODEC =
        new StreamCodec<>() {
            @Override
            public ParticleSpawnPayload decode(FriendlyByteBuf buf) {
                UUID pid = StreamCodecs.UUID_CODEC.decode(buf);
                ParticleStyle sty = ParticleStyle.values()[buf.readVarInt()];
                double x = buf.readDouble();
                double y = buf.readDouble();
                double z = buf.readDouble();
                float r = buf.readFloat();
                float g = buf.readFloat();
                float b = buf.readFloat();
                float a = buf.readFloat();
                float scale = buf.readFloat();
                int lifetime = buf.readVarInt();
                UUID gid = StreamCodecs.NULLABLE_UUID_CODEC.decode(buf);
                boolean glw = buf.readBoolean();
                return new ParticleSpawnPayload(pid, sty, x, y, z, r, g, b, a, scale, lifetime, gid, glw);
            }

            @Override
            public void encode(FriendlyByteBuf buf, ParticleSpawnPayload p) {
                StreamCodecs.UUID_CODEC.encode(buf, p.particleId);
                buf.writeVarInt(p.style.ordinal());
                buf.writeDouble(p.x);
                buf.writeDouble(p.y);
                buf.writeDouble(p.z);
                buf.writeFloat(p.r);
                buf.writeFloat(p.g);
                buf.writeFloat(p.b);
                buf.writeFloat(p.a);
                buf.writeFloat(p.scale);
                buf.writeVarInt(p.lifetime);
                StreamCodecs.NULLABLE_UUID_CODEC.encode(buf, p.groupId);
                buf.writeBoolean(p.glowing);
            }
        };

    @Override
    public Type<? extends CustomPacketPayload> type() {
        return TYPE;
    }

    public Vec3 position() {
        return new Vec3(x, y, z);
    }

    public Color color() {
        return Color.of(r, g, b, a);
    }
}

File: src/main/kotlin/work/nekow/particledrawing/core/network/ParticleSpawnPayload.kt
package work.nekow.particledrawing.core.network

import net.minecraft.network.FriendlyByteBuf
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.resources.Identifier
import net.minecraft.world.phys.Vec3
import work.nekow.particledrawing.api.Color
import java.util.UUID

/**
 * 粒子生成数据包，包含粒子的视觉和位置等所有属性。
 *
 * @param particleId 粒子唯一 ID
 * @param x/y/z 世界坐标
 * @param r/g/b/a RGBA 颜色分量
 * @param scale 渲染缩放
 * @param lifetime 存活 tick 数
 * @param groupId 所属组 ID，可为 null
 * @param glowing 是否发光
 * @param lightLevel 发光粒子向外发出的光照等级 (0-15)
 */
@Suppress("unused")
data class ParticleSpawnPayload(
    val particleId: UUID,
    val x: Double, val y: Double, val z: Double,
    val r: Float, val g: Float, val b: Float, val a: Float,
    val scale: Float,
    val lifetime: Int,
    val groupId: UUID?,
    val glowing: Boolean,
    val lightLevel: Int
) : CustomPacketPayload {

    companion object {
        @JvmField
        val TYPE = CustomPacketPayload.Type<ParticleSpawnPayload>(
            Identifier.fromNamespaceAndPath("particledrawing", "particle_spawn")
        )

        @JvmField
        val STREAM_CODEC: StreamCodec<FriendlyByteBuf, ParticleSpawnPayload> =
            object : StreamCodec<FriendlyByteBuf, ParticleSpawnPayload> {
                override fun decode(buf: FriendlyByteBuf): ParticleSpawnPayload {
                    val pid = StreamCodecs.UUID_CODEC.decode(buf)
                    val x = buf.readDouble()
                    val y = buf.readDouble()
                    val z = buf.readDouble()
                    val r = buf.readFloat()
                    val g = buf.readFloat()
                    val b = buf.readFloat()
                    val a = buf.readFloat()
                    val scale = buf.readFloat()
                    val lifetime = buf.readVarInt()
                    val gid = StreamCodecs.readNullableUUID(buf)
                    val glw = buf.readBoolean()
                    val light = buf.readVarInt()
                    return ParticleSpawnPayload(pid, x, y, z, r, g, b, a, scale, lifetime, gid, glw, light)
                }

                override fun encode(buf: FriendlyByteBuf, p: ParticleSpawnPayload) {
                    StreamCodecs.UUID_CODEC.encode(buf, p.particleId)
                    buf.writeDouble(p.x)
                    buf.writeDouble(p.y)
                    buf.writeDouble(p.z)
                    buf.writeFloat(p.r)
                    buf.writeFloat(p.g)
                    buf.writeFloat(p.b)
                    buf.writeFloat(p.a)
                    buf.writeFloat(p.scale)
                    buf.writeVarInt(p.lifetime)
                    StreamCodecs.writeNullableUUID(buf, p.groupId)
                    buf.writeBoolean(p.glowing)
                    buf.writeVarInt(p.lightLevel)
                }
            }
    }

    override fun type(): CustomPacketPayload.Type<out CustomPacketPayload> = TYPE

    fun position(): Vec3 = Vec3(x, y, z)

    fun color(): Color = Color.of(r, g, b, a)
}
